src/part-guide/concurrency.md

```diff
@@ -0,0 +1,175 @@
+
+Let's look at IO again. What happens when a thread blocks waiting for IO? In a system with threads, then the OS will pause the thread (it's just going to be waiting in any case) and wake it up again when the IO is complete[^busywait]. Depending on the scheduling algorithm, it might take some time after the IO completes until the OS wakes up the thread waiting for IO, since the OS might wait for other threads to get some work done. So now things are much more efficient: while one thread waits for IO, another thread (or more likely, many threads due to multitasking) can make progress. But, from the perspective of the thread doing IO, things are still sequential - it waits for the IO to finish before starting the next operation.
+
+A thread can also choose to pause itself by calling a `sleep` function, usually with a timeout. In this case the OS pauses the thread at the threads own request. Similar to pausing due to pre-emption or IO, the OS will wake the thread up again later (after the timeout) to continue execution.
+
+When an OS pauses one thread and starts another (for any reason), it is called *context switching*. The context being switched includes the registers, operating system records, and the contents of many caches. That's a non-trivial amount of work. Together with the transfer of control to the OS and back to a thread, and the costs of working with stale caches, context switching is an expensive operation.
+
+Finally, note that some hardware or OSs do not support processes or threads, this is more likely in the embedded world.
+
+[^proc-program]: from the user's perspective, a single program may include multiple processes, but from the OS's perspective each process is a separate program.
+[^shmem]: Some OSs do support sharing memory between processes, but using it requires special treatment and most memory is not shared.
+[^sched]: Exactly how the OS chooses which thread to run and for how long (and on which core), is a key part of scheduling. There are many options, both high-level strategies and options to configure those strategies. Making good choices here is crucial for good performance, but it is complicated and we won't dig into it here.
+[^busywait]: There's another option which is that the thread can *busy wait* by just spinning in a loop until the IO is finished. This is not very efficient since other threads won't get to run and is uncommon in most modern systems. You may come across it in the implementations of locks or in very simple embedded systems.
+
+
+## Async programming
```

Comment:
So, I think that's right, but I don't think this is the right place. This chapter is meant to be a high-level summary of the different kinds of concurrency, and in a practical way, what is similar and what is different. In particular I want to be clear and precise about the assumptions from threaded concurrency which carry over to async. The performance difference mentioned here is present more as a consequence of that, rather than because I'm trying to 'sell' async. I've made a conscious decision to avoid any syntax here and focus on tasks rather than futures or about how async is represented in the source code. So I don't think talking about join, etc. would fit well here.

I think that cancellation and timeouts will only murky the waters for this chapter - there are equivalents for threads, they're just not as ergonomic as for async, and they don't tend to feature in libraries, etc. However, there isn't a fundamental reason for why not, and I think getting into the distinctions and details wouldn't be appropriate here.

I do have a little section in the intro which more about motivation for choosing async, and I'll expand that to include more of these things.